qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -0,0 +1,324 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Translates gates to a target basis using a given equivalence library."""
+
+import time
+import logging
+
+from heapq import heappush, heappop
+from itertools import zip_longest
+from itertools import count as iter_count
+from collections import defaultdict
+
+import numpy as np
+
+from qiskit.circuit import Gate, ParameterVector, QuantumRegister
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+logger = logging.getLogger(__name__)
+
+
+class BasisTranslator(TransformationPass):
+    """Translates gates to a target basis using a given equivalence library."""
+
+    def __init__(self, equivalence_library, target_basis):
+        """Initialize a BasisTranslator instance.
+
+        Args:
+            equivalence_library (EquivalenceLibrary): The equivalence library
+                which will be used by the BasisTranslator pass. (Instructions in
+                this library will not be unrolled by this pass.)
+            target_basis (list[str]): Target basis names to unroll to, e.g. `['u3', 'cx']`.
+        """
+
+        super().__init__()
+
+        self._equiv_lib = equivalence_library
+        self._target_basis = target_basis
+
+    def run(self, dag):
+        """Translate an input DAGCircuit to the target basis.
+
+        Args:
+            dag (DAGCircuit): input dag
+
+        Raises:
+            TranspilerError: if the target basis cannot be reached
+
+        Returns:
+            DAGCircuit: translated circuit.
+        """
+
+        if self._target_basis is None:
+            return dag
+
+        # Names of instructions assumed to supported by any backend.
+        basic_instrs = ['measure', 'reset', 'barrier', 'snapshot']
+
+        target_basis = set(self._target_basis).union(basic_instrs)
+        source_basis = set((node.op.name, node.op.num_qubits)
+                           for node in dag.op_nodes())
```

Comment:
Yea, this was necessary so that we could handle gates with variable width like ms, but I agree, we need a more granular way to describe device basis gates than a gate name.